### ISSUE TITLE
Invalid property values in response on POST request using a nested parameter

### ISSUE THREAD

**Issue by nijikokun** [Kong/kong#3222]
### Summary

Making a `POST` request to create a `key-auth` plugin using _nested params_, creates a response where the configuration property `key_names` value becomes `[ 'function' ]`

### Steps To Reproduce

**Note** `HTTPie 0.9.8` will properly send the request required to replicate as it will send `null` values.

Make a POST request using JSON with a nested parameter:

```bash
http --debug :8001/plugins name=key-auth config.run_on_preflight=true config.key_names:=null
```

Get back a response which contains the invalid output:

```json
{
    "config": {
        "anonymous": "",
        "hide_credentials": false,
        "key_in_body": false,
        "key_names": [
            "function"
        ],
        "run_on_preflight": true
    },
    "created_at": 1518551355000,
    "enabled": true,
    "id": "6e07f3fe-5502-43e8-8e26-c1d9e3981855",
    "name": "key-auth"
}
```

Where:

```
        "key_names": [
            "function"
        ],
```

**Bonus Points**: Caching issue

Delete the Plugin you just created and run the following request:

```bash
http --debug :8001/plugins name=key-auth config.run_on_preflight=true
```

You will get the same result.

Delete the plugin again, and re-run the request above. You will get the proper result. Seems like there are retained values in memory.

### Additional Details & Logs

- Kong version (`0.11+`)
- Ubuntu

**Comment by nijikokun** [comment]
Another note, the plugin schema for `key-auth` states that the `key_names` property is _required_, the documentation states _optional_, and the request obviously went through without supplying an error about the missing field. Would you like another issue for this as well?

**Comment by bungle** [comment]
Some findings with the `function` there. I couldn't reproduce this yet, but it seems like the bug is introduced in the latest commit: https://github.com/Kong/kong/commit/cee785fdfcd96a72fa01fa9ff8c213c6f9c2a1e1 which adds `goto` and `::continue::` to code. I am not sure why it did work on my machine with or without that commit, but that is something to look for for further investigation.

**Comment by hishamhm** [comment]
I tried but wasn't able to reproduce this locally either. I downgraded my httpie to 0.9.8 (it reports HTTPie 0.9.8, Requests 2.18.4, Pygments 2.2.0, Python 2.7.12) and ran Kong on tags `master`, `0.12.1` and `0.11.2`.

**Comment by hishamhm** [comment]
Also tried it with curl to control the JSON explicitly:

```
curl -i -X POST -H "Content-Type:application/json" --url http://localhost:8001/plugins -d '{ "name": "key-auth", "config":{ "run_on_preflight": true, "key_names": null } }'
```

but got this:
```
HTTP/1.1 201 Created
Date: Wed, 14 Feb 2018 21:27:32 GMT
Content-Type: application/json; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
Access-Control-Allow-Origin: *
Server: kong/0.12.1

{"created_at":1518643652000,"config":{"key_in_body":false,"key_names":["apikey"],"anonymous":"","run_on_preflight":true,"hide_credentials":false},"id":"0b3fcc53-dc2e-445c-88d3-5836cdbfce35","enabled":true,"name":"key-auth"}
```

**Comment by hishamhm** [comment]
Sending an explicit value for `config.key_names` also works:

```
curl -i -X POST -H "Content-Type:application/json" --url http://localhost:8001/plugins -d '{ "name": "key-auth", "config":{ "run_on_preflight": true, "key_names": ["ovo"] } }'
```

It set it correctly:

```
HTTP/1.1 201 Created
Date: Wed, 14 Feb 2018 21:29:17 GMT
Content-Type: application/json; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
Access-Control-Allow-Origin: *
Server: kong/0.12.1

{"created_at":1518643757000,"config":{"key_in_body":false,"key_names":["ovo"],"anonymous":"","run_on_preflight":true,"hide_credentials":false},"id":"5e2326f2-e4c8-4e42-a52e-6321989aba16","enabled":true,"name":"key-auth"}
```

@nijikokun in which exact Kong version did you see the faulty behavior?

**Comment by nijikokun** [comment]
`0.11`, `0.10` variants.

**Comment by nijikokun** [comment]
Alright so, @hishamhm you were not going crazy, and I too thought I was going crazy, I could easily replicate that day and with fresh instances of docker, today, I could not replicate with fresh instances. It was driving me insane (not being able to replicate with httpie):

<img width="1661" alt="screen shot 2018-03-26 at 9 15 37 pm" src="https://user-images.githubusercontent.com/198276/37946572-f5bc9a12-313a-11e8-95c5-15e034f6d6fc.png">

Then... I decided, why not take exactly the entire request that caused the issue, and try that?

```
> curl -H "Content-Type: application/json" -X POST http://localhost:8001/plugins -d '{"name":"key-auth","api_id":null,"config.anonymous":null,"config.hide_credentials":false,"config.key_in_body":false,"config.key_names":null,"config.run_on_preflight":true}'

{"created_at":1522123978000,"config":{"key_in_body":false,"key_names":["function"],"anonymous":"","hide_credentials":false,"run_on_preflight":true},"id":"98b127c3-e1c5-4153-b26b-dca2f7cc0aa3","enabled":true,"name":"key-auth"}
```

I got the dreaded `['function']` result, here is where it gets even more interesting...

<img width="1667" alt="screen shot 2018-03-26 at 9 17 23 pm" src="https://user-images.githubusercontent.com/198276/37946626-3032eb10-313b-11e8-9f58-c328f31301c8.png">

Running the exact same command again... different result.

cc @thibaultcha

**Comment by hishamhm** [comment]
@nijikokun Great detective work! Reproduced on `master` with:

```
curl -i -X GET -H "Content-Type:application/json" --url http://localhost:8001/plugins/schema/key-auth
curl -i -X POST -H "Content-Type:application/json" --url http://localhost:8001/plugins -d '{ "name": "key-auth", "config":{ "run_on_preflight": true, "key_names": null } }'
```

**Comment by thibaultcha** [comment]
Merged #3348, expect a fix in the next minor!

**Comment by nijikokun** [comment]
Thanks @thibaultcha and @hishamhm ❤️
